File: src/agent_c_api_ui/agent_c_api/src/agent_c_api/api/rt/models/client_events.py
from typing import List, Optional, Any

from pydantic import Field

from agent_c.models import ChatSession
from agent_c.models.agent_config import CurrentAgentConfiguration, AgentCatalogEntry
from agent_c.models.events import BaseEvent
from agent_c.models.heygen import Avatar, HeygenAvatarSessionData, NewSessionRequest
from agent_c_api.core.voice.models import AvailableVoiceModel


class GetAgentsEvent(BaseEvent):
    """
    Event to request a list of available agents.

    This event does not require any additional data.
    """
    pass

class AgentListEvent(BaseEvent):
    """
    Event to send a list of available agents.

    Attributes:
        agents (list): A list of available agent keys.
    """
    agents: List[AgentCatalogEntry]

class SetAgentEvent(BaseEvent):
    """
    Event to set the agent for the current session.

    Attributes:
        agent_key (str): The key of the agent to set.
    """
    agent_key: str

class AgentConfigurationChangedEvent(BaseEvent):
    """
    Event to notify that the agent configuration has changed.

    Attributes:
        agent_config (CurrentAgentConfiguration): The new agent configuration.
    """
    agent_config: CurrentAgentConfiguration

class GetAvatarsEvent(BaseEvent):
    """
    Event to request a list of available avatars.

    This event does not require any additional data.
    """
    pass

class AvatarListEvent(BaseEvent):
    avatars: List[Avatar]

class SetAvatarSessionEvent(BaseEvent):
    """
    Contains the access token and avatar session ID to allow the server to drive the HeyGen avatar speech.
    """
    access_token: str
    avatar_session_id: str

class SetAvatarEvent(BaseEvent):
    """
    Event to set the avatar for the current session.

    Attributes:
        avatar_id (str): The ID of the avatar to set.
    """
    avatar_id: str
    quality: str = "medium"
    video_encoding: str = "H264"

class AvatarConnectionChangedEvent(BaseEvent):
    avatar_session_request: NewSessionRequest
    avatar_session: HeygenAvatarSessionData


class TextInputEvent(BaseEvent):
    """
    Event to send text input to the avatar.

    Attributes:
        text (str): The text input to send.
    """
    text: str
    file_ids: List[str] =  Field(list)

class ErrorEvent(BaseEvent):
    """
    Event to send an error message.

    Attributes:
        message (str): The error message to send.
    """
    message: str
    source: Optional[str] = None

class NewChatSessionEvent(BaseEvent):
    """
    Event to request a new chat session.

    This event does not require any additional data.
    """
    agent_key: Optional[str] = None

class ResumeChatSessionEvent(BaseEvent):
    """
    Event to request resuming a chat session.

    Attributes:
        session_id (str): The ID of the session to resume.
    """
    session_id: str

class ChatSessionChangedEvent(BaseEvent):
    """
    Event to notify that the chat session has changed.

    Attributes:
        chat_session (ChatSession): The updated chat session.
    """
    chat_session: ChatSession

class SetChatSessionNameEvent(BaseEvent):
    """
    Event to set the name of the current chat session.

    Attributes:
        session_name (str): The new name for the chat session.
    """
    session_name: str

class ChatSessionNameChangedEvent(BaseEvent):
    """
    Event to notify that the chat session name has changed.

    Attributes:
        session_name (str): The updated name of the chat session.
    """
    session_name: str

class SetSessionMetadataEvent(BaseEvent):
    """
    Event to set metadata for the current chat session.
    """
    meta: dict

class SessionMetadataChangedEvent(BaseEvent):
    """
    Event to notify that the chat session metadata has changed.

    Attributes:
        meta (dict): The updated metadata of the chat session.
    """
    meta: dict

class SetSessionMessagesEvent(BaseEvent):
    """
    Event to set messages for the current chat session.
    """
    messages: List[dict[str, Any]] = Field(default_factory=list, description="New list of messages in the session")

class SetAgentVoiceEvent(BaseEvent):
    """
    Event to set the voice for the current agent.

    Attributes:
        voice_id (str): The ID of the voice to set.
    """
    voice_id: str

class AgentVoiceChangedEvent(BaseEvent):
    """
    Event to notify that the agent voice has changed.

    Attributes:
        voice (AvailableVoiceModel): The updated voice of the agent.
    """
    voice: AvailableVoiceModel

class UserTurnStartEvent(BaseEvent):
    """
    Event to notify that the client we're accepting user input

    This event does not require any additional data.
    """
    pass

class UserTurnEndEvent(BaseEvent):
    """
    Event to notify that the client we're accepting user input

    This event does not require any additional data.
    """
    pass

File: src/realtime_client/packages/core/src/events/types/ClientEvents.ts
/**
 * Client-to-server event definitions for the Agent C Realtime API
 * These events are sent from the client to control the server
 */

import { Message } from './CommonTypes';

/**
 * Base interface for all client events
 */
export interface BaseClientEvent {
  type: string;
}

/**
 * Request list of available agents
 */
export interface GetAgentsEvent extends BaseClientEvent {
  type: 'get_agents';
}

/**
 * Set the active agent for the session
 */
export interface SetAgentEvent extends BaseClientEvent {
  type: 'set_agent';
  agent_key: string;
}

/**
 * Request list of available avatars
 */
export interface GetAvatarsEvent extends BaseClientEvent {
  type: 'get_avatars';
}

/**
 * Set the avatar for the current session
 */
export interface SetAvatarEvent extends BaseClientEvent {
  type: 'set_avatar';
  avatar_id: string;           // The ID of the avatar to set
  quality?: string;            // Optional quality setting (default: "auto")
  video_encoding?: string;     // Optional video encoding (default: "H265")
}

/**
 * Connect to an existing avatar session
 */
export interface SetAvatarSessionEvent extends BaseClientEvent {
  type: 'set_avatar_session';
  access_token: string;      // HeyGen access token for the session
  avatar_session_id: string;  // HeyGen avatar session ID
}

/**
 * Clear avatar session when HeyGen session ends
 */
export interface ClearAvatarSessionEvent extends BaseClientEvent {
  type: 'clear_avatar_session';
  session_id: string;
}

/**
 * Set the TTS voice for the current agent
 */
export interface SetAgentVoiceEvent extends BaseClientEvent {
  type: 'set_agent_voice';
  voice_id: string;
}

/**
 * Send text message to agent
 */
export interface TextInputEvent extends BaseClientEvent {
  type: 'text_input';
  text: string;
  file_ids?: string[];
}

/**
 * Create a new chat session
 */
export interface NewChatSessionEvent extends BaseClientEvent {
  type: 'new_chat_session';
  agent_key?: string;
}

/**
 * Resume an existing chat session
 */
export interface ResumeChatSessionEvent extends BaseClientEvent {
  type: 'resume_chat_session';
  session_id: string;
}

/**
 * Set the name/title for a chat session
 */
export interface SetChatSessionNameEvent extends BaseClientEvent {
  type: 'set_chat_session_name';
  session_name: string;
  session_id?: string; // Optional - if not set, applies to current session
}

/**
 * Update session metadata
 */
export interface SetSessionMetadataEvent extends BaseClientEvent {
  type: 'set_session_metadata';
  meta: Record<string, any>;
}

/**
 * Replace all messages in the current session
 */
export interface SetSessionMessagesEvent extends BaseClientEvent {
  type: 'set_session_messages';
  messages: Message[];
}

/**
 * Request paginated list of user sessions
 * Used for loading additional sessions beyond the initial login response
 */
export interface GetUserSessionsEvent extends BaseClientEvent {
  type: 'get_user_sessions';
  offset: number; // Starting offset for pagination (default 0)
  limit: number;  // Number of sessions to fetch (default 50)
}

/**
 * Request list of available TTS voices
 * Server responds with voice_list event
 */
export interface GetVoicesEvent extends BaseClientEvent {
  type: 'get_voices';
}

/**
 * Request tool catalog with available tools and schemas
 * Server responds with tool_catalog event
 */
export interface GetToolCatalogEvent extends BaseClientEvent {
  type: 'get_tool_catalog';
}

/**
 * Connection health check ping
 * Server responds with pong event
 */
export interface PingEvent extends BaseClientEvent {
  type: 'ping';
}

/**
 * Notify server that client is starting push-to-talk audio input
 */
export interface PushToTalkStartEvent extends BaseClientEvent {
  type: 'ptt_start';
}

/**
 * Notify server that client has finished push-to-talk audio input
 */
export interface PushToTalkEndEvent extends BaseClientEvent {
  type: 'ptt_end';
}

/**
 * Set the voice input mode for the session
 */
export interface SetVoiceInputModeEvent extends BaseClientEvent {
  type: 'set_voice_input_mode';
  mode: 'ptt' | 'vad';
}

/**
 * Delete a chat session belonging to the user
 */
export interface DeleteChatSessionEvent extends BaseClientEvent {
  type: 'delete_chat_session';
  session_id?: string; // Optional - if not set, deletes the current session
}

/**
 * Request to cancel the current agent response
 * Client sends this to notify it wants to cancel the current agent response
 */
export interface ClientWantsCancelEvent extends BaseClientEvent {
  type: 'client_wants_cancel';
}

/**
 * Union type of all client events
 */
export type ClientEvent =
  | GetAgentsEvent
  | SetAgentEvent
  | GetAvatarsEvent
  | SetAvatarEvent
  | SetAvatarSessionEvent
  | ClearAvatarSessionEvent
  | SetAgentVoiceEvent
  | TextInputEvent
  | NewChatSessionEvent
  | ResumeChatSessionEvent
  | SetChatSessionNameEvent
  | SetSessionMetadataEvent
  | SetSessionMessagesEvent
  | GetUserSessionsEvent
  | GetVoicesEvent
  | GetToolCatalogEvent
  | PingEvent
  | PushToTalkStartEvent
  | PushToTalkEndEvent
  | SetVoiceInputModeEvent
  | DeleteChatSessionEvent
  | ClientWantsCancelEvent;

/**
 * Type guard to check if an object is a client event
 */
export function isClientEvent(obj: unknown): obj is ClientEvent {
  return !!obj && typeof obj === 'object' && 'type' in obj && typeof (obj as any).type === 'string' && [
    'get_agents',
    'set_agent',
    'get_avatars',
    'set_avatar',
    'set_avatar_session',
    'clear_avatar_session',
    'set_agent_voice',
    'text_input',
    'new_chat_session',
    'resume_chat_session',
    'set_chat_session_name',
    'set_session_metadata',
    'set_session_messages',
    'get_user_sessions',
    'get_voices',
    'get_tool_catalog',
    'ping',
    'ptt_start',
    'ptt_end',
    'set_voice_input_mode',
    'delete_chat_session',
    'client_wants_cancel'
  ].includes((obj as any).type);
}